### PR TITLE
ExecuteQuery API ready for release

### PR DESCRIPTION
**Pull request by RichardIrons-neo4j**
README still needs to be updated to have samples using the new API.

### CODE REVIEW

File: Neo4j.Driver/Neo4j.Driver/IBookmarkManagerFactory.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-namespace Neo4j.Driver.Preview;
+namespace Neo4j.Driver;
 
 /// <summary>
 /// Preview: Subject to change.<br/> The <see cref="IBookmarkManagerFactory"/> interface is intended for
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/IBookmarkManagerFactory.cs]: can be removed

File: Neo4j.Driver/Neo4j.Driver/SessionConfig.cs [hunk 2]
```diff
@@ -281,8 +280,12 @@ internal SessionConfig Build()
         return _config;
     }
 
-    /// <summary>marked as internal until API is solidified.</summary>
-    internal SessionConfigBuilder WithBookmarkManager(IBookmarkManager bookmarkManager)
+    /// <summary>
+    /// <Sets the <see cref="IBookmarkManager"/> for maintaining bookmarks for the lifetime of the session.
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/SessionConfig.cs]:
```suggestion
    /// Sets the <see cref="IBookmarkManager"/> for maintaining bookmarks for the lifetime of the session.

```

File: Neo4j.Driver/Neo4j.Driver/Internal/Driver.cs [hunk 3]
```diff
@@ -145,30 +159,37 @@ public ValueTask DisposeAsync()
...
-        CancellationToken cancellationToken = default)
+        QueryConfig config,
+        Action<IRecord> streamProcessor,
+        CancellationToken cancellationToken
+    )
     {
-        return ExecuteQueryAsyncInternal(
+        async Task<int> Process(IAsyncEnumerable<IRecord> records)
+        {
+            await foreach (var record in records)
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Driver.cs]: await is not configured, can deadlock.

File: Neo4j.Driver/Neo4j.Driver/IBookmarkManager.cs [hunk 4]
```diff
@@ -18,14 +18,13 @@
...
 
 /// <summary>
-/// Preview: Subject to change.<br/> The <see cref="IBookmarkManager"/> interface is intended for
-/// implementation by classes that provide convenient interfacing with <see cref="Bookmarks"/> in both the driver and user
-/// code.
+/// The <see cref="IBookmarkManager"/> interface is intended for implementation by classes that provide convenient
+/// interfacing with <see cref="Bookmarks"/> in both the driver and user code.
 /// </summary>
-/// <see cref="PreviewExtensions.WithBookmarkManager"/>
+/// <see cref="ConfigBuilder."/>
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/IBookmarkManager.cs]:
```suggestion
/// <seealso cref="SessionConfigBuilder.WithBookmarkManager"/>
```

File: Neo4j.Driver/Neo4j.Driver/BookmarkManagerConfig.cs [hunk 5]
```diff
@@ -31,7 +31,7 @@ namespace Neo4j.Driver.Preview;
 /// </summary>
 /// <remarks>
 /// The default <see cref="IBookmarkManagerFactory"/> can be accessed from
-/// <see cref="GraphDatabase.BookmarkManagerFactory"/>.
+/// <see cref="Preview.GraphDatabase.BookmarkManagerFactory"/>.
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/BookmarkManagerConfig.cs]:
```suggestion
/// <see cref="GraphDatabase.BookmarkManagerFactory"/>.
```

File: Neo4j.Driver/Neo4j.Driver.Tests.TestBackend/Controller.cs [hunk 6]
```diff
@@ -98,7 +98,7 @@ public async Task Process(bool restartInitialState, Func<Exception, bool> loopCo
     {
         var restartConnection = restartInitialState;
 
-        Trace.WriteLine("Starting Controller.Process");
+        Trace.WriteLine("Starting Controller.Accumulate");
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver.Tests.TestBackend/Controller.cs]: seems accidental

**Review by AndyHeap-NeoTech — APPROVED** [review]
Looks good other than moving and renaming the FluentQueries folder as discussed.

File: Neo4j.Driver/Neo4j.Driver/Internal/Driver.cs [hunk 7]
```diff
@@ -223,16 +245,21 @@ await using (session.ConfigureAwait(false))
...
+
     private static Func<IResultCursor, CancellationToken, Task<EagerResult<TResult>>> TransformCursor<TResult>(
-        Func<IAsyncEnumerable<IRecord>, ValueTask<TResult>> streamProcessor)
+        Func<IAsyncEnumerable<IRecord>, Task<TResult>> streamProcessor)
     {
         async Task<EagerResult<TResult>> TransformCursorImpl(
             IResultCursor cursor,
             CancellationToken cancellationToken)
         {
             var processedStream = await streamProcessor(cursor);
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Driver.cs]:
```suggestion
            var processedStream = await streamProcessor(cursor).ConfigureAwait(false);
```

File: Neo4j.Driver/Neo4j.Driver/IDriver.cs [hunk 8]
```diff
@@ -93,4 +94,15 @@ public interface IDriver : IDisposable, IAsyncDisposable
     /// cluster support multi-databases, otherwise false.
     /// </returns>
     Task<bool> SupportsMultiDbAsync();
+
+    /// <summary>
+    /// Gets an <see cref="IExecutableQuery&lt;IRecord, IRecord&gt;"/> that can be used to configure and execute a query
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/IDriver.cs:99]: I am not sure if this will render correctly

Comment (thelonelyvulpes) [reply Neo4j.Driver/Neo4j.Driver/IDriver.cs:99]: can use `{` and `}` for types but looks good lically.

File: Neo4j.Driver/Neo4j.Driver/ExecuteQuery/IExecutableQuery.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    /// Specifies a method of reducing many items of type <see cref="TOut"/> into one instance of type
+    /// <see cref="TResult"/>, using one method to accumulate the values into a value of type <see cref="TAccumulate"/>
+    /// and another to turn the accumulated value into a result.
+    /// </summary>
+    /// <param name="seed">The initial value of the accumulating value.</param>
+    /// <param name="accumulate">A method that will accumulate each value of type <see cref="TOut"/> into
+    /// the accumulated value.</param>
+    /// <param name="selectResult">A method that will turn the accumulated value into a value of type
+    /// <see cref="TResult"/>.</param>
+    /// <typeparam name="TAccumulate"></typeparam>
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/ExecuteQuery/IExecutableQuery.cs:115]: missing docs

File: Neo4j.Driver/Neo4j.Driver/ExecuteQuery/IExecutableQuery.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    /// <summary>
+    /// Executes the query.
+    /// </summary>
+    /// <param name="token">A cancellation token that can be used to cancel the operation.</param>
+    /// <returns>An <see cref="EagerResult{TOut}"/>That contains the result of the query and
+    /// information about the execution.</returns>
+    Task<EagerResult<IReadOnlyList<TOut>>> ExecuteAsync(CancellationToken token = default);
+}
+
+public interface IReducedExecutableQuery<TOut>
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/ExecuteQuery/IExecutableQuery.cs:137]: missing docs

File: Neo4j.Driver/Neo4j.Driver/ExecuteQuery/EagerResult.cs [hunk 11]
```diff
@@ -15,11 +15,10 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-namespace Neo4j.Driver.Preview;
+namespace Neo4j.Driver;
 
 /// <summary>
-/// There is no guarantee that anything in Neo4j.Driver.Preview namespace will be in a next minor version.
-/// Complete result from a cypher query.
+/// Complete, materialised result from a cypher query.
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/ExecuteQuery/EagerResult.cs]:
```suggestion
/// Complete, materialized result from a cypher query.
```

File: Neo4j.Driver/Neo4j.Driver/ExecuteQuery/EagerResult.cs [hunk 12]
```diff
@@ -31,19 +30,16 @@ internal EagerResult(T result, IResultSummary summary, string[] keys)
...
-    /// There is no guarantee that anything in Neo4j.Driver.Preview namespace will be in a next minor version.
-    /// Least common set of fields in <see cref="Result"/>.
+    /// Smallest common set of fields in <see cref="Result"/>.
     /// </summary>
     public string[] Keys { get; init; }
 
     /// <summary>
-    /// There is no guarantee that anything in Neo4j.Driver.Preview namespace will be in a next minor version.
-    /// All Records from query.
+    /// The materialised result of the query.
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/ExecuteQuery/EagerResult.cs]:
```suggestion
    /// The materialized result of the query.
```

File: Neo4j.Driver/Neo4j.Driver/FluentQueries/EagerResult.cs [hunk 13]
```diff
@@ -31,19 +30,16 @@ internal EagerResult(T result, IResultSummary summary, string[] keys)
     }
 
     /// <summary>
-    /// There is no guarantee that anything in Neo4j.Driver.Preview namespace will be in a next minor version.
-    /// Least common set of fields in <see cref="Result"/>.
+    /// Smallest common set of fields in <see cref="Result"/>.
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/FluentQueries/EagerResult.cs]:
```suggestion
    /// Least common set of fields in <see cref="Result"/>.
```

File: Neo4j.Driver/Neo4j.Driver/FluentQueries/EagerResult.cs [hunk 14]
```diff
@@ -15,11 +15,10 @@
...
 
-namespace Neo4j.Driver.Preview;
+namespace Neo4j.Driver;
 
 /// <summary>
-/// There is no guarantee that anything in Neo4j.Driver.Preview namespace will be in a next minor version.
-/// Complete result from a cypher query.
+/// Complete, materialised result from a cypher query.
 /// </summary>
 public sealed class EagerResult<T>
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/FluentQueries/EagerResult.cs]: missing typeparam documentation

File: README.md [hunk 15]
```diff
@@ -61,54 +61,69 @@ PM> Install-Package Neo4j.Driver.Signed
 Connect to a Neo4j database
 
 ```csharp
-IDriver driver = GraphDatabase.Driver("neo4j://localhost:7687", AuthTokens.Basic("username", "pasSW0rd"));
-IAsyncSession session = driver.AsyncSession(o => o.WithDatabase("neo4j"));
-try
+using (var driver = GraphDatabase.Driver("bolt://localhost:7687", AuthTokens.Basic("neo4j", "password")))
```
Comment (thelonelyvulpes) [inline README.md]: I think using declaration would look even nicer.

File: README.md [hunk 16]
```diff
@@ -61,54 +61,69 @@ PM> Install-Package Neo4j.Driver.Signed
...
 ### Parsing Result Values
 
-#### Record Stream
+#### Query Execution Result
 
-A cypher execution result is comprised of a stream records followed by a result summary.
-The records inside the result are accessible via `FetchAsync` and `Current` methods on `IResultCursor`.
-Our recommended way to access these result records is to make use of methods provided by `ResultCursorExtensions` such
-as `SingleAsync`, `ToListAsync`, and `ForEachAsync`.
+The result of executing a query in the way shown above is an `EagerResult<T>`, where `T` is the type of data returned from the query. In the simplest case, this will be an `IReadOnlyList<IRecord>`, which is a fully materialised list of the records returned by the query. Other types of results can be used by using the fluent API exposed by the `IExecutableQuery` type.
```
Comment (thelonelyvulpes) [inline README.md]: I don't think we should remove the old `#### Record Stream` Section, and instead this new section should go before it

File: Neo4j.Driver/Neo4j.Driver/ExecuteQuery/EagerResult.cs (hunk 12 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/ExecuteQuery/EagerResult.cs]: Done

**Pull request merged**